### DISCUSSION TITLE
Built-in signals system

### DISCUSSION THREAD

**Discussion by hvuhsg** [fastapi/fastapi#6431] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get('/')
def some_endpoint():
    app.dispatch('user_created', user_data={'username': ..., 'password': ...})


@app.on_signal('user_created')
def send_notification(user_data):
    pass
```

### Description

I want an option to have built-in event system with custom events (not just, 'startup' and 'shutdown')

### Wanted Solution

Event system built-in and easy to use

### Wanted Code

```python
from fastapi import FastAPI

app = FastAPI()


@app.get('/')
def some_endpoint():
    app.dispatch('user_created', user_data={'username': ..., 'password': ...})


@app.on_signal('user_created')
def send_notification(user_data):
    pass
```

### Operating System

Linux

### FastAPI Version

0.70.0

### Python Version

3.8.0

**Comment by ArcLightSlavik** [comment, marked as answer]
Don't see why FastAPI needs to implement this, task queues basically accomplish this and there's already many popular libraries that do it. Startup and shutdown events are only there for server wide infrastructure like databases etc, not individual tasks.

**Comment by dstlny** [comment]
> ### First Check
> * [x]  I added a very descriptive title to this issue.
> * [x]  I used the GitHub search to find a similar issue and didn't find it.
> * [x]  I searched the FastAPI documentation, with the integrated search.
> * [x]  I already searched in Google "How to X in FastAPI" and didn't find any information.
> * [x]  I already read and followed all the tutorial in the docs and didn't find an answer.
> * [x]  I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
> * [x]  I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
> * [x]  I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
> 
> ### Commit to Help
> * [x]  I commit to help with one of those options 👆
> 
> ### Example Code
> ```python
> from fastapi import FastAPI
> 
> app = FastAPI()
> 
> 
> @app.get('/')
> def some_endpoint():
>     app.dispatch('user_created', user_data={'username': ..., 'password': ...})
> 
> 
> @app.on_signal('user_created')
> def send_notification(user_data):
>     pass
> ```
> 
> ### Description
> I want an option to have built-in event system with custom events (not just, 'startup' and 'shutdown')
> 
> ### Wanted Solution
> Event system built-in and easy to use
> 
> ### Wanted Code
> ```python
> from fastapi import FastAPI
> 
> app = FastAPI()
> 
> 
> @app.get('/')
> def some_endpoint():
>     app.dispatch('user_created', user_data={'username': ..., 'password': ...})
> 
> 
> @app.on_signal('user_created')
> def send_notification(user_data):
>     pass
> ```
> 
> ### Alternatives
> _No response_
> 
> ### Operating System
> Linux
> 
> ### Operating System Details
> _No response_
> 
> ### FastAPI Version
> 0.70.0
> 
> ### Python Version
> 3.8.0
> 
> ### Additional Context
> _No response_

I doubt this is going to get added, as it isn't the responsibility of FastAPI.

But, something like the following works - it just uses the background tasks functionality (and a lookup table)  to dispatch the signal after the request has ran:

```python
from fastapi import (
	FastAPI as BaseFastAPI,
	Request,
	BackgroundTasks,
	APIRouter
)


class FastAPI(BaseFastAPI):

	def __init__(self, *args, **kwargs):
		self._signals = {}
		super(BaseFastAPI, self).__init__(*args, **kwargs)

	def dispatch(
		self,
		background_tasks: BackgroundTasks,
		signal: str,
		*args,
		**kwargs
	):
		func_for_signal = self._signals.get(signal, None)

		if not func_for_signal:
			logging.error(f'No corresponding signal found for: {signal}')
			return
		
		background_tasks.add_task(func_for_signal, *args, **kwargs)
		
	def on_signal(
		self,
		signal: str
	):
		def decorator(func):
			func._signal = signal
			self._signals[signal] = func
			return func
		
		return decorator

app: FastAPI = FastAPI()


@app.get('/')
def some_endpoint(
	request: Request,
	background_tasks: BackgroundTasks
):
	app.dispatch(
		background_tasks=background_tasks,
		signal='user_created',
		user_data={'username': None, 'password': None}
	)

@app.on_signal('user_created')
def user_created_signal(user_data):
	print(f'Recieved signal: {user_created_signal._signal}')
```

Outputs:
```powershell
Invoke-RestMethod -Uri http://192.168.0.33:8000
```
```
INFO:     2021-11-04 10:08:31 - uvicorn.access:send:461 - 192.168.0.33:60449 - "GET / HTTP/1.1" 200
Recieved signal: user_created
````

Be aware though, the BackgroundTasks functionality tends to block if it's a super long running task - and if it is super long running, you'd be better off using `asyncio.create_task` or similar.

**Comment by tiangolo** [comment]
Thanks for the help here everyone! 👏  🙇 

Thanks for reporting back and closing the issue 👍

And yep, this is something that would not really belong in FastAPI, probably could be done with the integrated background tasks, or with Celery, RQ, ARQ, or another similar background job library.

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
